src/librustc_typeck/collect.rs

```diff
@@ -195,9 +293,12 @@ impl AstConv<'tcx> for ItemCtxt<'tcx> {
         None
     }
 
-    fn ty_infer(&self, _: Option<&ty::GenericParamDef>, span: Span) -> Ty<'tcx> {
-        bad_placeholder_type(self.tcx(), span).emit();
+    fn allow_ty_infer(&self) -> bool {
+        false
+    }
 
+    fn ty_infer(&self, _: Option<&ty::GenericParamDef>, span: Span) -> Ty<'tcx> {
+        self.tcx().sess.delay_span_bug(span, "bad placeholder type, but no error was emitted");
```

Comment:
the "but no error was emitted" part is always true for `delay_span_bug` and thus not necessary to add to the message